### PR TITLE
Move advanced span creation options to new class to simply API

### LINKED ISSUE #297: Span creation options: for advanced overloads use span creation options
following suggestion here https://github.com/open-telemetry/opentelemetry-dotnet/pull/268#issuecomment-540716028. or we are going to drown in different startSpan overloads.

We should have basic span creation args available through overloads without builder/options
- name
- kind
- parent

advanced things should be populated through options. 
- sampling hint
- timestamp
- links
- lazy link factory
- any future advanced things 

Users create options (and allocate) when they want to provide advanced options. Simple `StartSpan` overloads that don't need advanced options, don't allocate this options object and pass null instance instead.

### PR DESCRIPTION
**Pull request by lmolkova**
Fixes #297 

Good things: 
- half of Start methods on tracer
- Extension story: when we need to introduce new creation options - we can put them into `SpanCreationOptions` without breaking anyone. We know we'll need it soon for #273 and sampling hint

Bad things:
- `SpanCreationOptions` is a class and all advanced scenarios require allocation even in noop case. Making it a struct may be an option, but breaks extensibility story and hard to implement (all fields have to set in constructor). 

Why I decided to still do this despite bad things: I tried to make friendly and efficient APIs to start scoped/active spans and drowned in existing Start overloads.

I also kept links enum for `StartFromActivity`
- timestamp comes with Activity - never `SpoanCreationOptions`
- `StartFromActivity` works with auto-collectors, it implies there instrumentation and it's not a noop case (i.e. having Func<Links> makes not much sense)

### CODE REVIEW

**Comment by pcwiese** [comment]
I love how you've reduced the API surface of ITracer. Looks much better to my eyes! I suppose you could add some object pooling/reuse for the SpanCreationOptions class if you are concerned with allocations...

File: src/OpenTelemetry.Abstractions/Trace/ITracer.cs [hunk 1]
```diff
@@ -50,101 +51,45 @@ public interface ITracer
...
-        /// </summary>
-        /// <param name="operationName">Span name.</param>
-        /// <param name="parent">Parent for new span.</param>
-        /// <param name="kind">Kind.</param>
-        /// <param name="startTimestamp">Start timestamp.</param>
-        /// <param name="linksGetter">Func to retrieve the Links collection.</param>
+        /// <param name="options">Advanced span creation options.</param>
         /// <returns>Span instance.</returns>
-        ISpan StartSpan(string operationName, ISpan parent, SpanKind kind, DateTimeOffset startTimestamp, Func<IEnumerable<Link>> linksGetter);
+        ISpan StartRootSpan(string operationName, SpanKind kind, SpanCreationOptions options);
```
Comment (pcwiese) [inline src/OpenTelemetry.Abstractions/Trace/ITracer.cs]: Could these two methods become extensions also?

Comment (lmolkova) [reply src/OpenTelemetry.Abstractions/Trace/ITracer.cs]: StartSpan can and will in a different PR. StartRootspan has special meaning - ignore any implicit parents which could be expressed as a flag or in the method name, so I don't see a good option on how to move it to extensions. Let me know if you have any solution in mind

**Comment by lmolkova** [comment]
> I love how you've reduced the API surface of ITracer. Looks much better to my eyes! I suppose you could add some object pooling/reuse for the SpanCreationOptions class if you are concerned with allocations...

Thanks, you wait until I break this again in the other PR :)

**Pull request merged**